main/openssh/sshd.initd

```diff
@@ -1,100 +1,45 @@
 #!/sbin/openrc-run
-# Copyright 1999-2015 Gentoo Foundation
-# Distributed under the terms of the GNU General Public License v2
-# $Header: /var/cvsroot/gentoo-x86/net-misc/openssh/files/sshd.rc6.4,v 1.5 2015/05/04 02:56:25 vapier Exp $
 
 description="OpenBSD Secure Shell server"
-description_checkconfig="Verify configuration file"
 description_reload="Reload configuration"
+description_checkconfig="Verify configuration file"
 
-extra_commands="checkconfig"
-extra_started_commands="reload"
-
-: ${SSHD_CONFDIR:=/etc/ssh}
-: ${SSHD_CONFIG:=${SSHD_CONFDIR}/sshd_config}
-: ${SSHD_PIDFILE:=/run/${SVCNAME}.pid}
-: ${SSHD_BINARY:=/usr/sbin/sshd}
+cfgfile=${SSHD_CONFIG:-${SSHD_CONFDIR:-/etc/ssh}/sshd_config}
+pidfile=${SSHD_PIDFILE:-/run/$RC_SVCNAME.pid}
+command=${SSHD_BINARY:-/usr/sbin/sshd}
+command_args="-D $SSHD_OPTS"
+command_background=yes
+extra_commands=checkconfig
+extra_started_commands=reload
+required_files=$cfgfile
 
 depend() {
 	use logger dns
-	if [ "${rc_need+set}" = "set" ] ; then
-		: # Do nothing, the user has explicitly set rc_need
-	else
-		local x warn_addr
-		for x in $(awk '/^ListenAddress/{ print $2 }' "$SSHD_CONFIG" 2>/dev/null) ; do
-			case "${x}" in
-				0.0.0.0|0.0.0.0:*) ;;
-				::|\[::\]*) ;;
-				*) warn_addr="${warn_addr} ${x}" ;;
-			esac
-		done
-		if [ -n "${warn_addr}" ] ; then
-			need net
-			ewarn "You are binding an interface in ListenAddress statement in your sshd_config!"
-			ewarn "You must add rc_need=\"net.FOO\" to your /etc/conf.d/sshd"
-			ewarn "where FOO is the interface(s) providing the following address(es):"
-			ewarn "${warn_addr}"
-		fi
-	fi
 }
 
-checkconfig() {
-	if [ ! -d /var/empty ] ; then
-		mkdir -p /var/empty || return 1
+start_pre() {
+	if [ "$RC_CMD" = "start" ]; then
+		checkpath -d /var/empty ${pidfile%/*}
+		yesno $SSHD_DISABLE_KEYGEN || ssh-keygen -A || return 1
 	fi
-
-	if [ ! -e "${SSHD_CONFIG}" ] ; then
-		eerror "You need an ${SSHD_CONFIG} file to run sshd"
-		eerror "There is a sample file in /usr/share/doc/openssh"
-		return 1
-	fi
-
-	if ! yesno "${SSHD_DISABLE_KEYGEN}"; then
-		ssh-keygen -A || return 1
-	fi
-
-	[ "${SSHD_PIDFILE}" != "/run/sshd.pid" ] \
-		&& SSHD_OPTS="${SSHD_OPTS} -o PidFile=${SSHD_PIDFILE}"
-	[ "${SSHD_CONFIG}" != "/etc/ssh/sshd_config" ] \
-		&& SSHD_OPTS="${SSHD_OPTS} -f ${SSHD_CONFIG}"
-
-	"${SSHD_BINARY}" -t ${SSHD_OPTS} || return 1
+	[ "$cfgfile" = "/etc/ssh/sshd_config" ] || command_args="$command_args -f $cfgfile"
+	$command -t $command_args
 }
 
-start() {
-	checkconfig || return 1
-
-	ebegin "Starting ${SVCNAME}"
-	start-stop-daemon --start --exec "${SSHD_BINARY}" \
-	    --pidfile "${SSHD_PIDFILE}" \
-	    -- ${SSHD_OPTS}
-	eend $?
+stop_post() {
+	local pids
+	[ "$RC_RUNLEVEL" = "shutdown" ] && pids=$(pgrep ${command##*/})
+	[ -z "$pids" ] || kill -TERM $pids >/dev/null 2>&1
 }
 
-stop() {
-	if [ "${RC_CMD}" = "restart" ] ; then
-		checkconfig || return 1
```

Comment:
I'd like to keep the `checkconfig` on `restart` before stopping the service. Maybe we can do:

````sh
stop_pre() {
        if [ "${RC_CMD}" = "restart" ] ; then
            checkconfig || return 1
        fi
}
````

or:
````sh
stop_pre() {
        [ "${RC_CMD}" != "restart" ] && return 0
        checkconfig
}
````

Former is more readable.